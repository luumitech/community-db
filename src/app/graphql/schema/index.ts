import { builder } from '../builder';

import './access';
import './community';
import './geocode';
import './offset-pagination';
import './payment';
import './property';
import './subscription';
import './user';

export const schema = builder.toSchema();
