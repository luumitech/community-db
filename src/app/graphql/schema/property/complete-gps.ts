import { Property, Role } from '@prisma/client';
import { GraphQLError } from 'graphql';
import { builder } from '~/graphql/builder';
import { MessageType } from '~/graphql/pubsub';
import { GeoapifyApi } from '~/lib/geoapify-api';
import prisma from '~/lib/prisma';
import { verifyAccess } from '../access/util';
import { getCommunityEntry } from '../community/util';

const PropertyCompleteGpsInput = builder.inputType('PropertyCompleteGpsInput', {
  fields: (t) => ({
    communityId: t.string({ required: true }),
  }),
});

interface PropertyCompleteGpsPayload {
  propertyList: Property[];
}

const propertyCompleteGpsPayloadRef = builder
  .objectRef<PropertyCompleteGpsPayload>('PropertyCompleteGpsPayload')
  .implement({
    fields: (t) => ({
      propertyList: t.prismaField({
        type: ['Property'],
        resolve: (query, result, args, ctx) => result.propertyList,
      }),
    }),
  });

builder.mutationField('propertyCompleteGps', (t) =>
  t.field({
    type: propertyCompleteGpsPayloadRef,
    description:
      'For property without GPS information, complete its GPS information using the address',
    args: {
      input: t.arg({ type: PropertyCompleteGpsInput, required: true }),
    },
    resolve: async (_parent, args, ctx) => {
      const { user, pubSub } = ctx;
      const { input } = args;
      const { communityId: shortId } = input;

      // Make sure user has permission to modify
      await verifyAccess(user, { shortId }, [Role.ADMIN, Role.EDITOR]);

      const community = await getCommunityEntry(user, shortId, {
        select: { id: true, propertyList: true },
      });

      // Get list of properties missing GPS information
      const propertyList = community.propertyList.filter((property) => {
        return !property.lat || !property.lon;
      });

      // Get geocode information for each address and update database with
      // information
      const api = await GeoapifyApi.fromConfig();
      const addressList = propertyList.map((property) => property.address);
      const results = await api.batchGeocode.searchFreeForm(addressList);

      const [...updatedPropertyList] = await prisma.$transaction([
        ...propertyList.map((property, idx) => {
          const geocodeResult = results[idx];
          // It's possible that lat/lon cannot be found
          return prisma.property.update({
            where: { id: property.id },
            data: {
              updatedBy: { connect: { email: user.email } },
              lat: geocodeResult.lat?.toString() ?? null,
              lon: geocodeResult.lon?.toString() ?? null,
            },
          });
        }),
      ]);

      // broadcast modification to property(s)
      updatedPropertyList.forEach((property) => {
        pubSub.publish(`community/${shortId}/property`, {
          broadcasterId: user.email,
          messageType: MessageType.UPDATED,
          property,
        });
      });

      return {
        propertyList: updatedPropertyList,
      };
    },
  })
);
