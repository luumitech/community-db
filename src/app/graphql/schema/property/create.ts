import { Role } from '@prisma/client';
import { GraphQLError } from 'graphql';
import { builder } from '~/graphql/builder';
import prisma from '~/lib/prisma';
import { verifyAccess } from '../access/util';
import { getCommunityEntry } from '../community/util';
import { getSubscriptionEntry } from '../payment/util';

const PropertyCreateInput = builder.inputType('PropertyCreateInput', {
  fields: (t) => ({
    communityId: t.string({
      description: 'community short ID',
      required: true,
    }),
    address: t.string({ description: 'property address', required: true }),
    streetNo: t.int({ description: 'street number' }),
    streetName: t.string({ description: 'street name', required: true }),
    postalCode: t.string({ description: 'postal code' }),
    city: t.string({ description: 'city' }),
    country: t.string({ description: 'country' }),
    lat: t.string({ description: 'GPS latitude' }),
    lon: t.string({ description: 'GPS longitude' }),
  }),
});

builder.mutationField('propertyCreate', (t) =>
  t.prismaField({
    type: 'Property',
    args: {
      input: t.arg({ type: PropertyCreateInput, required: true }),
    },
    resolve: async (query, _parent, args, ctx) => {
      const { user } = ctx;
      const { communityId, ...input } = args.input;

      // Make sure user has permission to create
      await verifyAccess(user, { shortId: communityId }, [Role.ADMIN]);

      // Cannot connect to community with shortId, so retrieve
      // the document directly
      const community = await getCommunityEntry(user, communityId, {
        select: { id: true, owner: true },
      });

      // Check community owner's subscription status to determine
      // limitation
      const existingSub = await getSubscriptionEntry(community.owner);
      const { propertyLimit } = existingSub;
      if (propertyLimit != null) {
        const propertyCount = await prisma.property.count({
          where: { community: { id: community.id } },
        });
        if (propertyCount >= propertyLimit) {
          throw new GraphQLError(
            `This community can at most have ${propertyLimit} properties.`
          );
        }
      }

      const entry = await prisma.property.create({
        ...query,
        data: {
          ...input,
          community: {
            connect: { id: community.id },
          },
        },
      });
      return entry;
    },
  })
);
