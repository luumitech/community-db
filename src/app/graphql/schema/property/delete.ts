import { Role } from '@prisma/client';
import { builder } from '~/graphql/builder';
import { MutationType } from '~/graphql/pubsub';
import prisma from '~/lib/prisma';
import { verifyAccess } from '../access/util';
import { getPropertyEntry } from './util';

const propertyDeletePayloadRef = builder
  .objectRef<{ id: string }>('PropertyDeletePayload')
  .implement({
    fields: (t) => ({
      id: t.exposeID('id', {}),
    }),
  });

builder.mutationField('propertyDelete', (t) =>
  t.field({
    type: propertyDeletePayloadRef,
    args: {
      id: t.arg.string({ required: true }),
    },
    resolve: async (_parent, args, ctx) => {
      const { user, pubSub } = await ctx;

      const propertyShortId = args.id;

      const entry = await getPropertyEntry(user, propertyShortId, {
        select: {
          id: true,
          updatedAt: true,
          community: {
            select: {
              shortId: true,
            },
          },
        },
      });

      // Make sure user has permission to delete
      await verifyAccess(user, { shortId: entry.community.shortId }, [
        Role.ADMIN,
      ]);

      const property = await prisma.property.delete({
        where: { id: entry.id },
      });

      // broadcast deletion to community
      pubSub.publish(
        `community/${entry.community.shortId}/property/${propertyShortId}`,
        {
          broadcasterId: user.email,
          mutationType: MutationType.DELETED,
          property,
        }
      );

      return property;
    },
  })
);
