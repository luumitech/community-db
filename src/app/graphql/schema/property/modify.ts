import { ApolloError } from '@apollo/client';
import { Community, Prisma, Property, Role } from '@prisma/client';
import { GraphQLError } from 'graphql';
import { builder } from '~/graphql/builder';
import { MutationType } from '~/graphql/pubsub';
import { extractYearRange } from '~/lib/lcra-community/import/year-range-util';
import prisma from '~/lib/prisma';
import { verifyAccess } from '../access/util';
import { UpdateInput } from '../common';
import {
  communityMinMaxYearUpdateArgs,
  getCommunityEntry,
} from '../community/util';
import { getPropertyEntry, propertyListFindManyArgs } from './util';

const OccupantInput = builder.inputType('OccupantInput', {
  fields: (t) => ({
    firstName: t.string(),
    lastName: t.string(),
    optOut: t.boolean(),
    email: t.string(),
    cell: t.string(),
    work: t.string(),
    home: t.string(),
  }),
});

const EventInput = builder.inputType('EventInput', {
  fields: (t) => ({
    eventName: t.string({ required: true }),
    eventDate: t.string(),
    ticket: t.int(),
  }),
});

const MembershipInput = builder.inputType('MembershipInput', {
  fields: (t) => ({
    year: t.int({ required: true }),
    eventAttendedList: t.field({ type: [EventInput] }),
    paymentMethod: t.string(),
    price: t.string(),
  }),
});

const PropertyModifyInput = builder.inputType('PropertyModifyInput', {
  fields: (t) => ({
    self: t.field({ type: UpdateInput, required: true }),
    address: t.string(),
    streetNo: t.int(),
    streetName: t.string(),
    postalCode: t.string(),
    notes: t.string(),
    occupantList: t.field({ type: [OccupantInput] }),
    membershipList: t.field({ type: [MembershipInput] }),
  }),
});

interface PropertyModifyPayload {
  community: Community;
  property: Property;
}

const propertyModifyPayloadRef = builder
  .objectRef<PropertyModifyPayload>('PropertyModifyPayload')
  .implement({
    fields: (t) => ({
      community: t.prismaField({
        type: 'Community',
        resolve: (query, result, args, ctx) => result.community,
      }),
      property: t.prismaField({
        type: 'Property',
        resolve: (query, result, args, ctx) => result.property,
      }),
    }),
  });

builder.mutationField('propertyModify', (t) =>
  t.field({
    type: propertyModifyPayloadRef,
    args: {
      input: t.arg({ type: PropertyModifyInput, required: true }),
    },
    resolve: async (_parent, args, ctx) => {
      const { user, pubSub } = await ctx;
      const { self, ...input } = args.input;
      const shortId = self.id;

      const entry = await getPropertyEntry(user, shortId, {
        select: {
          id: true,
          updatedAt: true,
          community: {
            select: {
              id: true,
              shortId: true,
              minYear: true,
              maxYear: true,
            },
          },
        },
      });
      if (entry.updatedAt.toISOString() !== self.updatedAt) {
        throw new GraphQLError(
          `Attempting to update a stale property ${shortId}, please refresh browser.`
        );
      }

      // Make sure user has permission to modify
      await verifyAccess(user, { shortId: entry.community.shortId }, [
        Role.ADMIN,
        Role.EDITOR,
      ]);

      // Check if community min/max year needs to be updated
      const yearRange = extractYearRange([input]);
      const communityUpdateDataArgs = communityMinMaxYearUpdateArgs(
        entry.community,
        yearRange.minYear,
        yearRange.maxYear
      );

      const [updatedCommunity, updatedProperty] = await prisma.$transaction([
        // Update community, if required
        communityUpdateDataArgs == null
          ? prisma.community.findUniqueOrThrow({
              where: { id: entry.community.id },
            })
          : prisma.community.update({
              where: { id: entry.community.id },
              data: communityUpdateDataArgs,
            }),

        // Update property
        prisma.property.update({
          where: {
            id: entry.id,
          },
          // @ts-expect-error: composite types like 'occupantList'
          // is allowed to be undefined
          data: {
            updatedBy: { connect: { email: user.email } },
            ...input,
          },
        }),
      ]);

      // broadcast modification to property
      pubSub.publish(`community/${entry.community.shortId}/property`, {
        broadcasterId: user.email,
        mutationType: MutationType.UPDATED,
        property: updatedProperty,
      });
      return {
        community: updatedCommunity,
        property: updatedProperty,
      };
    },
  })
);

export const PropertyFilterInput = builder.inputType('PropertyFilterInput', {
  fields: (t) => ({
    searchText: t.string({
      description: 'Match against property address/first name/last name',
    }),
    memberYear: t.int({
      description: 'Only property who is a member of the given year',
    }),
    nonMemberYear: t.int({
      description: 'Only property who is NOT a member of the given year',
    }),
    memberEvent: t.string({
      description: 'Only property who attended this event',
    }),
  }),
});

const BatchMembershipInput = builder.inputType('BatchMembershipInput', {
  fields: (t) => ({
    year: t.int({ required: true }),
    eventAttended: t.field({ type: EventInput, required: true }),
    paymentMethod: t.string({ required: true }),
    price: t.string(),
  }),
});

const BatchPropertyModifyInput = builder.inputType('BatchPropertyModifyInput', {
  fields: (t) => ({
    communityId: t.string({ required: true }),
    filter: t.field({ type: PropertyFilterInput }),
    membership: t.field({ type: BatchMembershipInput, required: true }),
  }),
});

interface BatchPropertyModifyPayload {
  community: Community;
  propertyList: Property[];
}

const batchPropertyModifyPayloadRef = builder
  .objectRef<BatchPropertyModifyPayload>('BatchPropertyModifyPayload')
  .implement({
    fields: (t) => ({
      community: t.prismaField({
        type: 'Community',
        resolve: (query, result, args, ctx) => result.community,
      }),
      propertyList: t.prismaField({
        type: ['Property'],
        resolve: (query, result, args, ctx) => result.propertyList,
      }),
    }),
  });

builder.mutationField('batchPropertyModify', (t) =>
  t.field({
    type: batchPropertyModifyPayloadRef,
    args: {
      input: t.arg({ type: BatchPropertyModifyInput, required: true }),
    },
    resolve: async (_parent, args, ctx) => {
      const { user, pubSub } = await ctx;
      const { input } = args;
      const { communityId: shortId, filter } = input;
      const { eventAttended } = input.membership;

      // Make sure user has permission to modify
      await verifyAccess(user, { shortId }, [Role.ADMIN, Role.EDITOR]);

      const community = await getCommunityEntry(user, shortId, {
        select: { id: true, minYear: true, maxYear: true },
      });

      const findManyArgs = await propertyListFindManyArgs(community.id, filter);
      const propertyList = await prisma.property.findMany(findManyArgs);

      // Modify the propertyList in memory, and then write them to
      // database afterwards
      propertyList.forEach((property) => {
        const membership = property.membershipList.find(
          (entry) => entry.year === input.membership.year
        );
        if (membership) {
          const event = membership.eventAttendedList.find(
            (entry) => entry.eventName === eventAttended.eventName
          );
          if (!event) {
            membership.eventAttendedList.push({
              eventName: eventAttended.eventName,
              eventDate: eventAttended.eventDate
                ? new Date(eventAttended.eventDate)
                : null,
              ticket: null,
            });
            // Non empty event list require payment Method
            if (membership.eventAttendedList.length === 1) {
              membership.paymentMethod = input.membership.paymentMethod;
            }
          }
        } else {
          property.membershipList.unshift({
            year: input.membership.year,
            paymentMethod: input.membership.paymentMethod,
            paymentDeposited: null,
            price: input.membership.price ?? null,
            eventAttendedList: [
              {
                eventName: eventAttended.eventName,
                eventDate: eventAttended.eventDate
                  ? new Date(eventAttended.eventDate)
                  : null,
                ticket: null,
              },
            ],
          });
        }
      });

      const communityUpdateDataArgs = communityMinMaxYearUpdateArgs(
        community,
        input.membership.year
      );

      const [updatedCommunity, ...updatedPropertyList] =
        await prisma.$transaction([
          // Update minYear/maxYear when appropriate
          communityUpdateDataArgs == null
            ? prisma.community.findUniqueOrThrow({
                where: { id: community.id },
              })
            : prisma.community.update({
                where: { id: community.id },
                data: communityUpdateDataArgs,
              }),
          ...propertyList.map((property) =>
            prisma.property.update({
              where: { id: property.id },
              data: {
                updatedBy: { connect: { email: user.email } },
                membershipList: property.membershipList,
              },
            })
          ),
        ]);

      // broadcast modification to property(s)
      updatedPropertyList.forEach((property) => {
        pubSub.publish(`community/${shortId}/property`, {
          broadcasterId: user.email,
          mutationType: MutationType.UPDATED,
          property,
        });
      });

      return {
        community: updatedCommunity,
        propertyList: updatedPropertyList,
      };
    },
  })
);
