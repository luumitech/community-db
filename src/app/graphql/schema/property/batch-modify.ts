import { Community, Property, Role } from '@prisma/client';
import { GraphQLError } from 'graphql';
import { builder } from '~/graphql/builder';
import { MessageType } from '~/graphql/pubsub';
import prisma from '~/lib/prisma';
import { verifyAccess } from '../access/util';
import { UpdateInput } from '../common';
import {
  communityMinMaxYearUpdateArgs,
  getCommunityEntry,
} from '../community/util';
import { EventInput } from './modify';
import {
  findOrAddEvent,
  mapEventEntry,
  propertyListFindManyArgs,
} from './util';

export const PropertyFilterInput = builder.inputType('PropertyFilterInput', {
  fields: (t) => ({
    searchText: t.string({
      description: 'Match against property address/first name/last name',
    }),
    memberYear: t.int({
      description: 'Only property who is a member of the given year',
    }),
    nonMemberYear: t.int({
      description: 'Only property who is NOT a member of the given year',
    }),
    memberEvent: t.string({
      description: 'Only property who attended this event',
    }),
    withGps: t.boolean({
      description: 'Only property with GPS coordinates',
    }),
  }),
});

const BatchMembershipInput = builder.inputType('BatchMembershipInput', {
  fields: (t) => ({
    year: t.int({ required: true }),
    eventAttended: t.field({ type: EventInput, required: true }),
    paymentMethod: t.string({ required: true }),
    price: t.string(),
  }),
});

const BatchPropertyModifyInput = builder.inputType('BatchPropertyModifyInput', {
  fields: (t) => ({
    self: t.field({ type: UpdateInput, required: true }),
    filter: t.field({ type: PropertyFilterInput }),
    membership: t.field({ type: BatchMembershipInput, required: true }),
  }),
});

interface BatchPropertyModifyPayload {
  community: Community;
  propertyList: Property[];
}

const batchPropertyModifyPayloadRef = builder
  .objectRef<BatchPropertyModifyPayload>('BatchPropertyModifyPayload')
  .implement({
    fields: (t) => ({
      community: t.prismaField({
        type: 'Community',
        resolve: (query, result, args, ctx) => result.community,
      }),
      propertyList: t.prismaField({
        type: ['Property'],
        resolve: (query, result, args, ctx) => result.propertyList,
      }),
    }),
  });

builder.mutationField('batchPropertyModify', (t) =>
  t.field({
    type: batchPropertyModifyPayloadRef,
    args: {
      input: t.arg({ type: BatchPropertyModifyInput, required: true }),
    },
    resolve: async (_parent, args, ctx) => {
      const { user, pubSub } = ctx;
      const { input } = args;
      const { self, filter } = input;
      const { eventAttended } = input.membership;
      const shortId = self.id;

      // Make sure user has permission to modify
      await verifyAccess(user, { shortId }, [Role.ADMIN, Role.EDITOR]);

      const community = await getCommunityEntry(user, shortId, {
        select: { id: true, minYear: true, maxYear: true },
      });

      const findManyArgs = await propertyListFindManyArgs(community.id, filter);
      const propertyList = await prisma.property.findMany(findManyArgs);

      // Modify the propertyList in memory, and then write them to
      // database afterwards
      propertyList.forEach(({ membershipList }) => {
        const result = findOrAddEvent(
          membershipList,
          input.membership.year,
          eventAttended.eventName
        );
        const membership = membershipList[result.membershipIdx];
        if (result.isNewEvent) {
          membership.eventAttendedList[result.eventIdx] =
            mapEventEntry(eventAttended);
        }
        if (result.isNewMember) {
          // Membership Fee will only be applied to properties
          // that do not have an existing membership
          membership.price = input.membership.price ?? null;
          membership.paymentMethod = input.membership.paymentMethod;
        } else if (membership.price == null) {
          // If a property has a Membership Fee entry, but does not have
          // Price information specified, the record will be updated with
          // the new Price information.
          membership.price = input.membership.price ?? null;
        }
      });

      const communityUpdateDataArgs = communityMinMaxYearUpdateArgs(
        community,
        input.membership.year
      );

      const [updatedCommunity, ...updatedPropertyList] =
        await prisma.$transaction([
          // Update minYear/maxYear when appropriate
          communityUpdateDataArgs == null
            ? prisma.community.findUniqueOrThrow({
                where: { id: community.id },
              })
            : prisma.community.update({
                where: { id: community.id },
                data: communityUpdateDataArgs,
              }),
          ...propertyList.map((property) =>
            prisma.property.update({
              where: { id: property.id },
              data: {
                updatedBy: { connect: { email: user.email } },
                membershipList: property.membershipList,
              },
            })
          ),
        ]);

      // broadcast modification to property(s)
      updatedPropertyList.forEach((property) => {
        pubSub.publish(`community/${shortId}/property`, {
          broadcasterId: user.email,
          messageType: MessageType.UPDATED,
          property,
        });
      });

      return {
        community: updatedCommunity,
        propertyList: updatedPropertyList,
      };
    },
  })
);
