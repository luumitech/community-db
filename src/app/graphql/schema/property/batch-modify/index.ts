import { Job } from '@hokify/agenda';
import { Role } from '@prisma/client';
import { GraphQLError } from 'graphql';
import { builder } from '~/graphql/builder';
import { verifyAccess } from '~/graphql/schema/access/util';
import { UpdateInput } from '~/graphql/schema/common';
import { jobPayloadRef } from '~/graphql/schema/job/object';
import { type ContextUser } from '~/lib/context-user';
import { JobHandler } from '~/lib/job-handler';
import { EventInput } from '../modify';
import { BatchModify } from './batch-modify';

export const PropertyFilterInput = builder.inputType('PropertyFilterInput', {
  fields: (t) => ({
    searchText: t.string({
      description: 'Match against property address/first name/last name',
    }),
    memberYear: t.int({
      description: 'Only property who is a member of the given year',
    }),
    nonMemberYear: t.int({
      description: 'Only property who is NOT a member of the given year',
    }),
    memberEvent: t.string({
      description: 'Only property who attended this event',
    }),
    withGps: t.boolean({
      description:
        'If true, properties with GPS.  If false, properties without GPS',
    }),
  }),
});

const BatchMembershipInput = builder.inputType('BatchMembershipInput', {
  fields: (t) => ({
    year: t.int({ required: true }),
    eventAttended: t.field({ type: EventInput, required: true }),
    paymentMethod: t.string({ required: true }),
    price: t.string(),
  }),
});

const batchModifyMethodRef = builder.enumType('BatchModifyMethod', {
  values: ['ADD_EVENT', 'ADD_GPS'] as const,
});

const BatchGpsInput = builder.inputType('BatchGpsInput', {
  fields: (t) => ({
    city: t.string(),
    province: t.string(),
    country: t.string(),
  }),
});

const BatchPropertyModifyInput = builder.inputType('BatchPropertyModifyInput', {
  fields: (t) => ({
    self: t.field({ type: UpdateInput, required: true }),
    method: t.field({ type: batchModifyMethodRef, required: true }),
    filter: t.field({ type: PropertyFilterInput }),
    membership: t.field({ type: BatchMembershipInput }),
    gps: t.field({ type: BatchGpsInput }),
  }),
});

builder.mutationField('batchPropertyModify', (t) =>
  t.field({
    type: jobPayloadRef,
    args: {
      input: t.arg({ type: BatchPropertyModifyInput, required: true }),
    },
    resolve: async (_parent, args, ctx) => {
      const { user, pubSub } = ctx;
      const { input } = args;
      const shortId = input.self.id;

      // Make sure user has permission to modify
      await verifyAccess(user, { shortId }, [Role.ADMIN, Role.EDITOR]);

      const agenda = await JobHandler.init();

      const job = await agenda.start<BatchPropertyModifyJobArg>(
        'batchPropertyModify',
        { user, input }
      );

      return job;
    },
  })
);

export type BatchPropertyModifyInput =
  typeof BatchPropertyModifyInput.$inferInput;
export interface BatchPropertyModifyJobArg {
  user: ContextUser;
  input: BatchPropertyModifyInput;
}

export async function batchPropertyModifyTask(
  job: Job<BatchPropertyModifyJobArg>
) {
  const batchModify = BatchModify.fromJob(job);

  const propertyList = await batchModify.modify();
  /**
   * Would be nice to store this information in agenda queue output, so we can
   * display number of properties modified by the batch command. But I don't
   * think this is currently possible
   */
  return propertyList.length;
}
