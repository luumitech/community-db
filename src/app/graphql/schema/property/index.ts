import './batch-modify';
import './complete-gps';
import './create';
import './delete';
import './modify';
import './object';
import './register-event';
