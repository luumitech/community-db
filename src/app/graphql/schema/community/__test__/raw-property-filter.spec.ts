import path from 'path';
import { graphql } from '~/graphql/generated';
import { TestUtil } from '~/graphql/test-util';
import { propertyListPrismaQuery } from '~/lib/prisma-raw-query/property';

const document = graphql(/* GraphQL */ `
  query RawPropertyFilterSpec_RawPropertyList($query: JSONObject!) {
    userCurrent {
      accessList {
        community {
          rawPropertyList(query: $query) {
            address
          }
        }
      }
    }
  }
`);

describe('Raw property query', () => {
  const testUtil = new TestUtil();

  beforeAll(async () => {
    await testUtil.initialize();
    await testUtil.database.seed(
      path.join(process.cwd(), '__fixtures__', 'simple-lcra.xlsx')
    );
  });

  afterAll(async () => {
    await testUtil.terminate();
  });

  test('Verify properties that have not been renewed', async () => {
    const query = propertyListPrismaQuery({
      nonMemberYear: 2024,
      memberYear: 2023,
    });
    const result = await testUtil.graphql.executeSingle({
      document,
      variables: { query },
    });
    expect(result.data).toMatchSnapshot();
  });

  test('Verify properties that are members', async () => {
    const query = propertyListPrismaQuery({
      memberYear: 2022,
    });
    const result = await testUtil.graphql.executeSingle({
      document,
      variables: { query },
    });
    expect(result.data).toMatchSnapshot();
  });

  test('Verify properties that are NOT members', async () => {
    const query = propertyListPrismaQuery({
      nonMemberYear: 2022,
    });
    const result = await testUtil.graphql.executeSingle({
      document,
      variables: { query },
    });
    expect(result.data).toMatchSnapshot();
  });
});
