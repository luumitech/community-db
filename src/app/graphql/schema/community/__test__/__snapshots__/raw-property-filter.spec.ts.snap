// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Raw property query Verify properties that are NOT members 1`] = `
{
  "__typename": "Query",
  "userCurrent": {
    "__typename": "User",
    "accessList": [
      {
        "__typename": "Access",
        "community": {
          "__typename": "Community",
          "rawPropertyList": [
            {
              "__typename": "Property",
              "address": "123 Adventure Drive",
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`Raw property query Verify properties that are members 1`] = `
{
  "__typename": "Query",
  "userCurrent": {
    "__typename": "User",
    "accessList": [
      {
        "__typename": "Access",
        "community": {
          "__typename": "Community",
          "rawPropertyList": [
            {
              "__typename": "Property",
              "address": "99 Fortune Drive",
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`Raw property query Verify properties that have not been renewed 1`] = `
{
  "__typename": "Query",
  "userCurrent": {
    "__typename": "User",
    "accessList": [
      {
        "__typename": "Access",
        "community": {
          "__typename": "Community",
          "rawPropertyList": [
            {
              "__typename": "Property",
              "address": "123 Adventure Drive",
            },
            {
              "__typename": "Property",
              "address": "99 Fortune Drive",
            },
          ],
        },
      },
    ],
  },
}
`;
